docs/how-to/debug-in-cluster.md

```diff
@@ -0,0 +1,59 @@
+
+![Location of the Kubernetes plugin in the plugin bar (screen left), with the Clusters>cluster>Workloads>Pods views expanded out to show a pod named "my-service", overlaid with a dropdown box, with "Attach Visual Studio Code" highlighted](../images/debugging-kubernetes.jpg)
+The Kubernetes plugin can be found in the plugin bar. Expanding the Clusters>`cluster`>Workloads>Pods views, your service should be visible. Right Click>Attach Visual Studio Code will initiate connecting to the workspace in the cluster. Select your service container from the top menu when prompted.
+
+After the connection to the cluster has been established, it may be necessary to open the workspace folder by clicking the Explorer option in the plugin bar, the repository will be mounted at `/workspaces/<service name>`, equivalent to when working with a local devcontainer.
+
+Starting your service with the command in the container definition starts it on the node, with access to Kubernetes resources, however it is also now possible to run with or attach a debugger, potentially configured to autoReload code, or to start and stop the service rapidly to implement prospective changes.
+
+After you are happy with the changes, commit them and release a new version of your container. Changes will otherwise not be persisted across container restarts. Your git and ssh config will be mounted inside the devcontainer while connected and for containers on github, the remote `origin` will be configured to use ssh.
+
+## Debugging containers that run as non-root
+For containers running in the Diamond Kubernetes infrastructure that run as a specific uid (e.g. if mounting the filesystem), it is required to use a sidecar container to provide name resolution from Diamond's LDAP infrastructure and to mount a home directory to house vscode plugins. 
+
+A sidecar for the Debian-based Python image this template uses is published as a container from this repository, the version should match the version of the python-copier-template you are using, to ensure compatibility with the underlying container infrastructure.
+
+```yaml
+- name: debug-account-sync
+  image: ghcr.io/diamondlightsource/python-copier-template/account-sync:<version>
+  volumeMounts:
+  # The nslcd socket will be shared between the service and the sidecar
+  - mountPath: /var/run/nslcd
+    name: nslcd
+```
+
+The following changes/additions to your `values.yaml` may be required to connect vscode when using the sidecar.
+It is recommended to set the `HOME` environment variable on your container to be debugged to the same value used in the volume below.
+
+```yaml
+volumes:
+- name: home  # Required for vscode to start and install plugins
+  hostPath:
+    path: /home/<fedid>
```

Comment:
Using just /home/ had issues with "too many levels of symlinks". Not sure if it's just my home dir.

Reply:
1. which helm chart are we using here? 
2. do you really need to mount a real home folder? can't we just set HOME to point at folder in the nslcd PVC?

Reply:
I've seen the symlinks problem when mounting automount folders and there is a workaround for this podman but not sure if that is available to K8S.

I'm not sure you will always be able to mount your own home dir though? If you are not running in your personal namespace then you won't be allowed to run with your own user id. A beamline namespace service account or ixxdetector account would not be allowed to mount your home directory (at least not writeable). Maybe I've missed something here?

I'm afraid I'm away until next Tues now - will discuss on my return.

Reply:
Yes AFAIK (and I should update the docs if this isn't clear) the home has to be the home of the user that the pod is running as.

i.e.
```yaml
podSecurityContext:
  runAsUser: <some uid>
  runAsGroup: <some gid>

volumeMounts:
- mountPath: /home/<username of the uid>/
  name: home
```

Reply:
OK, I see. I'm pretty sure that shared namespace service accounts don't have a home directory.

I feel that making a PVC and using that to as a home directory for the vscode config would be better. Each developer could make their own if they really wanted and keep it for debugging in the cluster, whichever account the service being debugged requires to run as.

Reply:
I feel like that's a slippery slope to re-inventing all of our problems with home directory quotas into the file system.

I think I've justified to myself moving it into the shared volume: the devcontainer is defined from the devcontainer json, not from anything in your home dir. It can be thrown away when the devcontainer isn't required and recreated when the next person needs it.